File: octave/gamevalue.m
% Copyright © 2012 Bart Massey
% [This program is licensed under the "MIT License"]
% Please see the file COPYING in the source
% distribution of this software for license terms.

% Perfect Tic-Tac-Toe player in Matlab / Octave


% Returns the value of the game to the
% side on move if the game is over, or
% -2 if the game is still in progress.
function v = gamevalue(onmove, S)
  % first scan for wins
  for side = [-onmove, onmove]
    v = side * onmove;
    % scan for diagonal
    n = 0;
    for d = 1:3
      if S(d, d) == side
        n = n + 1;
      end
    end
    if n == 3
      return
    end
    % scan for opposite diagonal
    n = 0;
    for d = 1:3
      if S(d, 4 - d) == side
        n = n + 1;
      end
    end  
    if n == 3
      return
    end
    % scan for rows
    for r = 1:3
      n = 0;
      for c = 1:3
        if S(r, c) == side
          n = n + 1;
        end
      end  
      if n == 3
        return
      end
    end
    % scan for cols
    for c = 1:3
      n = 0;
      for r = 1:3
        if S(r, c) == side
          n = n + 1;
        end
      end
      if n == 3
        return
      end
    end
  end
  % scan for blanks
  v = -2;   % game not over
  for r = 1:3
    for c = 1:3
      if S(r, c) == 0
        return
      end
    end
  end
  % game is a draw
  v = 0;
end

File: go/src/gamevalue/gamevalue.go
// Perfect Tic-Tac-Toe player in Go
// Copyright © 2017 Bart Massey
// [This program is licensed under the "MIT License"]
// Please see the file COPYING in the source
// distribution of this software for license terms.

// Returns the value of the game to the
// side on move if the game is over, or
// -2 if the game is still in progress.

package gamevalue

func GameValue(onmove int, board *[3][3]int) int {
	var v int
	// first scan for wins
	var side = -1
	for s := 0; s < 2; s++ {
		side = -side
		v = side * onmove
		// scan for diagonal
		n := 0
		for d := 0; d < 3; d++ {
			if board[d][d] == side {
				n++
			}
		}
		if n == 3 {
			return v
		}
		// scan for opposite diagonal
		n = 0
		for d := 0; d < 3; d++ {
			if board[d][2 - d] == side {
				n++
			}
		}
		if n == 3 {
			return v
		}
		// scan for rows
		for r := 0; r < 3; r++ {
			n := 0
			for c := 0; c < 3; c++ {
				if board[r][c] == side {
					n++
				}
			}
			if n == 3 {
				return v
			}
		}
		// scan for columns
		for c := 0; c < 3; c++ {
			n := 0
			for r := 0; r < 3; r++ {
				if board[r][c] == side {
					n++
				}
			}
			if n == 3 {
				return v
			}
		}
	}
	// scan for blanks
	for r := 0; r < 3; r++ {
		for c := 0; c < 3; c++ {
			if board[r][c] == 0 {
				// game not over
				return -2
			}
		}
	}
	// game is a draw
	return 0
}
